pkg/transport/fetchidentity.go

```diff
@@ -0,0 +1,101 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package transport
+
+import (
+	"context"
+	"net"
+
+	"github.com/zeebo/errs"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"storj.io/storj/pkg/identity"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/peertls"
+)
+
+// TODO: check whether this would be problematic with concurrency or should we thread the authinfo through context?
+type handshakeCapture struct {
+	credentials.TransportCredentials
+	AuthInfo credentials.AuthInfo
+}
+
+// ClientHandshake does the authentication handshake specified by the corresponding
+// authentication protocol on rawConn for clients. It returns the authenticated
+// connection and the corresponding auth information about the connection.
+func (capture *handshakeCapture) ClientHandshake(ctx context.Context, s string, conn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	conn, auth, err := capture.TransportCredentials.ClientHandshake(ctx, s, conn)
+	capture.AuthInfo = auth
+	return conn, auth, err
+}
+
+// ServerHandshake does the authentication handshake for servers. It returns
+// the authenticated connection and the corresponding auth information about
+// the connection.
+func (capture *handshakeCapture) ServerHandshake(conn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	conn, auth, err := capture.TransportCredentials.ServerHandshake(conn)
+	capture.AuthInfo = auth
+	return conn, auth, err
```

Comment:
nit - does the error need to be wrapped?

Reply:
here, no, because it's wrapping and we don't know how the error is handled internally, by wrapping we might lose important interface methods.

Reply:
We have to find out a way to let the callers know with errors (mostly methods that return errors) are fine to be inspected like this one and which ones not.

Reply:
I suspect that the error warpping might hide essential interfaces, such as `interface { Temporary() bool }`. https://godoc.org/google.golang.org/grpc#FailOnNonTempDialError, however I haven't checked the errs implementation for it.

Reply:
sure, I was only saying that we should find away that the callers can know which methods return original errors so they can do type assertions.